spec/formatting.md

```diff
@@ -219,15 +222,66 @@ the following steps are taken:
    as long as reasonable precautions are taken to keep the function interface
    simple and minimal, and avoid introducing potential security vulnerabilities.
 
-   As implementations MAY allow custom functions to be defined by users,
-   their access to the _formatting context_ SHOULD be minimal and read-only,
-   and their execution time SHOULD be limited.
+   Implementations MAY expose _options_ in an implementation-defined _namespace_
+   in addition to or superceding those found in the default registry.
+
+   An implementation MAY define its own functions.
+   An implementation MAY allow custom functions to be defined by users.
+
+   Function access to the _formatting context_ MUST be minimal and read-only,
+   and execution time SHOULD be limited.
+   Implementation-defined _functions_ SHOULD use an implementation-defined _namespace_.
+   User-defined _functions_ SHOULD be permitted to define the _namespace_ to use
+   in name resolution.
 
 5. If the call succeeds,
    resolve the value of the _expression_ as the result of that function call.
    If the call fails or does not return a valid value,
    emit a Resolution error and use a _fallback value_ for the _expression_.
 
+### Name Resolution
+
+<dfn title="resolve the name">**_Name resolution_**</dfn> is the determination of
+the _name-part_ and, where present, the _namespace_ of the _name_ for a given _function_
+or _option_.
+
+To resolve the _name_ of an item, the following steps are taken:
+
+1. Let the _name_ be the string to be processed, less any preceeding
+   sigils or markers.
+   For example, in the _expression_ `{:function}`, the string `function`
+   is the _name_ string.
+2. Determine if _name_ contains the character U+003A COLON `:`:
+   - If _name_ contains a colon, let _namespace_ be the
+      substring of _name_ up to (but not including)
+      the colon and _name-part_ be the substring of _name_ following the colon;
+   - Else let _namespace_ be an empty value (null, void, or empty string)
+     and _name-part_ be the _name_.
+3. If the _name-part_ is empty, return a Syntax error.
+4. Return _namespace_ and _name-part_.
+
+Implementations are not required to support the installation or resolution
+of different namespaces.
+How namespaces are defined, provisioned, or handled is also implementation defined.
+It is important to note that the _namespace_ is not considered part of the _name_
+after resolution. 
+The _namespace_ MAY affect which function is called.
+
+Example:
+> Suppose the ICU implementation of MessageFormat added a custom option `skeleton`
+> to the `:datetime` function.
+> Users of that implementation could invoke the datetime formatter with an
+> expression similar to:
```

Comment:
```suggestion
> pattern similar to:
```
(What follows is a pattern; alternately, you could leave this as-is and change the example to `{$date :datetime icu:skeleton=yMMMd}`)